runtime/src/iree-amd-aie/driver/xrt/direct_command_buffer.cc

```diff
@@ -0,0 +1,446 @@
+  target_offset += iree_hal_buffer_byte_offset(target_buffer);
+
+  xrt::bo* source_device_buffer = iree_hal_xrt_buffer_handle(
+      iree_hal_buffer_allocated_buffer(source_buffer));
+
+  void* source_device_buffer_ptr = source_device_buffer->map();
+  source_offset += iree_hal_buffer_byte_offset(source_buffer);
+  uint8_t* dst = (uint8_t*)target_device_buffer_ptr + target_offset;
+  uint8_t* src = (uint8_t*)source_device_buffer_ptr + source_offset;
+  memcpy(dst, src, length);
+
+  IREE_TRACE_ZONE_END(z0);
+  return iree_ok_status();
+}
+
+static iree_status_t iree_hal_xrt_direct_command_buffer_collective(
+    iree_hal_command_buffer_t* base_command_buffer, iree_hal_channel_t* channel,
+    iree_hal_collective_op_t op, uint32_t param,
+    iree_hal_buffer_binding_t send_binding,
+    iree_hal_buffer_binding_t recv_binding, iree_device_size_t element_count) {
+  return iree_make_status(IREE_STATUS_UNIMPLEMENTED,
+                          "collectives not yet supported");
+}
+
+static iree_status_t iree_hal_xrt_direct_command_buffer_push_constants(
+    iree_hal_command_buffer_t* base_command_buffer,
+    iree_hal_pipeline_layout_t* pipeline_layout, iree_host_size_t offset,
+    const void* values, iree_host_size_t values_length) {
+  return iree_make_status(IREE_STATUS_UNIMPLEMENTED,
+                          "push constants not yet supported");
+}
+
+static iree_status_t iree_hal_xrt_direct_command_buffer_push_descriptor_set(
+    iree_hal_command_buffer_t* base_command_buffer,
+    iree_hal_pipeline_layout_t* pipeline_layout, uint32_t set,
+    iree_host_size_t binding_count,
+    const iree_hal_descriptor_set_binding_t* bindings) {
+  if (binding_count > IREE_HAL_XRT_MAX_DESCRIPTOR_SET_BINDING_COUNT) {
+    return iree_make_status(
+        IREE_STATUS_RESOURCE_EXHAUSTED,
+        "exceeded available binding slots for push "
+        "descriptor set #%" PRIu32 "; requested %" PRIhsz " vs. maximal %d",
+        set, binding_count, IREE_HAL_XRT_MAX_DESCRIPTOR_SET_BINDING_COUNT);
+  }
+
+  iree_hal_xrt_direct_command_buffer_t* command_buffer =
+      iree_hal_xrt_direct_command_buffer_cast(base_command_buffer);
+  IREE_TRACE_ZONE_BEGIN(z0);
+
+  xrt::bo** current_bindings = command_buffer->descriptor_sets[set].bindings;
+  for (iree_host_size_t i = 0; i < binding_count; i++) {
+    const iree_hal_descriptor_set_binding_t* binding = &bindings[i];
+    if (binding->buffer) {
+      continue;
+    }
+    IREE_TRACE_ZONE_END(z0);
+    return iree_make_status(IREE_STATUS_UNIMPLEMENTED,
+                            "unimplemented null buffer in push descriptor set");
+  }
+  for (iree_host_size_t i = 0; i < binding_count; i++) {
+    const iree_hal_descriptor_set_binding_t* binding = &bindings[i];
+    IREE_RETURN_AND_END_ZONE_IF_ERROR(
+        z0, iree_hal_resource_set_insert(command_buffer->resource_set, 1,
+                                         &binding->buffer));
+    std::unique_ptr<xrt::bo> sub_buffer;
+    try {
```

Comment:
To provide some more details. I think we can extend the current `iree_hal_xrt_buffer_t` to contain three more fields: a `xrt::bo` `base_buffer`, a `iree_device_size_t` `offset` and `length`. Then add a new `iree_hal_xrt_allocator_create_sub_buffer` API in direct_allocator.h, which does all the `make_unique` logic here and be adjacent to what we have in the `iree_hal_xrt_allocator_allocate_buffer` function. It would internally call another new `iree_hal_xrt_sub_buffer_wrap` API in `xrt_buffer.h`. `iree_hal_xrt_sub_buffer_wrap` should take parameters like `xrt::bo` base buffer, `iree_device_size_t` `offest` and `length` which are needed for sub buffers, while keep the current `iree_hal_xrt_buffer_wrap` as-is (or maybe renaming it into `iree_Hal_xrt_base_buffer_wrap` to differentiate. Need to adjust various xrt buffer api impl to accomodate the new fields too.